### ISSUE TITLE
Liquid containers accepting nonsensical, non-liquid contents

### ISSUE THREAD

**Issue by DatMeg** [CleverRaven/Cataclysm-DDA#40950]
# Describe the bug

Despite the implied usage, and shape of a Canteen, it appears capable of holding a wide variety of non-liquid objects that shouldn't (logically) fit within it, even if the volume of the object would fit. It's certainly a sneaky place to store your pistols, but I'm not sure how you'd get them in it, or back out for that matter.

# Steps To Reproduce

1. spawn a Canteen (and maybe even wear it)
2. Find, or spawn an object with a volume less than, or equal to that of the Canteen (1.5L)
3. Somehow find a way to stick that rollmat, gun, or even that holster with a gun in it... into it.

# Expected behavior

I realize that placement is based almost entirely on volume, and one could even argue that there are some small things that would be perfectly fine in a Canteen (like ammunition). However there are some things that probably shouldn't go in some places. A rollmat that squeezes into a canteen or gallon jug probably isn't a very good surface to sleep on.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1909
- Game Version: 0.E-2605-g14dcde8 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda]
]

**Comment by Lamandus** [comment]
placement is not only based on volume anymore, but also maximum lengh and "item-length" as well. So either the canteen is missing the max_lengh, or the object you put in has the wrong lengh.

**Comment by EvgenijM86** [comment]
If we are talking about realism then you should be able to insert some items that have volume larger than container mouth. For example: something shaped like a cylinder.

**Comment by KorGgenT** [comment]
we are not fully doing aperture sizes, this is the compromise i came up with in order to not give every single item in the game three dimensions. there will be some things that would theoretically fit into other things, but the impact of the game will be relatively low.

**Comment by DatMeg** [comment]
> 
> 
> we are not fully doing aperture sizes, this is the compromise i came up with in order to not give every single item in the game three dimensions. there will be some things that would theoretically fit into other things, but the impact of the game will be relatively low.

I wasn't really expecting us to do the apertures, it was more an amusing addition than expectation, because of the image of stuffing a loaded holster or an entire rollmat down into a canteen or jug was... amusing.

**Comment by solifuge** [comment]
Since the issue seems to not actually be closed yet, I'll add that some liquid containers have no way to accept liquid contents. I can insert a smartphone into a hipflask, but I can't pour antiseptic or water into it. Unload/Reload doesn't recognize some liquid containers either (hipflask).

**Comment by solifuge** [comment]
I've been trying to do some container behavior testing, but it seems to have corrupted my save, making further testing difficult. I unloaded batteries from a bunch of devices I had equipped, carried, and stored on surrounding shelves/furniture. Once I'd done that, attempting to grab items from one of those shelves/tables (or attempting to target that furniture in the Advanced Inventory, or attempting to move everything in the [A]ll Tile mode) resulted in a segfault. Same if I attempted to scroll the Advanced Inventory to display those batteries, I think (crashes before the menu updates, so it's hard to say). Also, hitting [r] to reload while near those shelves crashes with an arithmetic error (I assume it's something about trying to access those same batteries).

I have crash logs from both, if they're of any use. I was also getting errors on load, about an item's pocket (a holster) not being able to fit "items of that type," and errors that atomic batteries were loaded with charge, but cannot have any.
